types/key/node.go

```diff
@@ -0,0 +1,244 @@
+}
+
+func ParseNodePrivateUntyped(raw mem.RO) (NodePrivate, error) {
+	var ret NodePrivate
+	if err := parseHex(ret.k[:], raw, mem.B(nil)); err != nil {
+		return NodePrivate{}, err
+	}
+	return ret, nil
+}
+
+// IsZero reports whether k is the zero value.
+func (k NodePrivate) IsZero() bool {
+	return k.Equal(NodePrivate{})
+}
+
+// Equal reports whether k and other are the same key.
+func (k NodePrivate) Equal(other NodePrivate) bool {
+	return subtle.ConstantTimeCompare(k.k[:], other.k[:]) == 1
+}
+
+// Public returns the NodePublic for k.
+// Panics if NodePrivate is zero.
+func (k NodePrivate) Public() NodePublic {
+	if k.IsZero() {
+		panic("can't take the public key of a zero NodePrivate")
+	}
+	var ret NodePublic
+	curve25519.ScalarBaseMult(&ret.k, &k.k)
+	return ret
+}
+
+// MarshalText implements encoding.TextMarshaler.
+func (k NodePrivate) MarshalText() ([]byte, error) {
+	return toHex(k.k[:], nodePrivateHexPrefix), nil
+}
+
+// MarshalText implements encoding.TextUnmarshaler.
+func (k *NodePrivate) UnmarshalText(b []byte) error {
+	return parseHex(k.k[:], mem.B(b), mem.S(nodePrivateHexPrefix))
+}
+
+// SealTo wraps cleartext into a NaCl box (see
+// golang.org/x/crypto/nacl) to p, authenticated from k, using a
+// random nonce.
+//
+// The returned ciphertext is a 24-byte nonce concatenated with the
+// box value.
+func (k NodePrivate) SealTo(p NodePublic, cleartext []byte) (ciphertext []byte) {
+	if k.IsZero() || p.IsZero() {
+		panic("can't seal with zero keys")
+	}
+	var nonce [24]byte
+	rand(nonce[:])
+	return box.Seal(nonce[:], cleartext, &nonce, &p.k, &k.k)
+}
+
+// OpenFrom opens the NaCl box ciphertext, which must be a value
+// created by SealTo, and returns the inner cleartext if ciphertext is
+// a valid box from p to k.
+func (k NodePrivate) OpenFrom(p NodePublic, ciphertext []byte) (cleartext []byte, ok bool) {
+	if k.IsZero() || p.IsZero() {
+		panic("can't open with zero keys")
+	}
+	if len(ciphertext) < 24 {
+		return nil, false
+	}
+	var nonce [24]byte
+	copy(nonce[:], ciphertext)
+	return box.Open(nil, ciphertext[len(nonce):], &nonce, &p.k, &k.k)
+}
+
+func (k NodePrivate) UntypedHexString() string {
+	return hex.EncodeToString(k.k[:])
+}
+
+// NodePublic is the public portion of a a NodePrivate.
+type NodePublic struct {
+	k [32]byte
+}
+
+// ParseNodePublicUntyped parses an untyped 64-character hex value
+// as a NodePublic.
+//
+// Deprecated: this function is risky to use, because it cannot verify
+// that the hex string was intended to be a NodePublic. This can
+// lead to accidentally decoding one type of key as another. For new
+// uses that don't require backwards compatibility with the untyped
+// string format, please use MarshalText/UnmarshalText.
+func ParseNodePublicUntyped(raw mem.RO) (NodePublic, error) {
+	var ret NodePublic
+	if err := parseHex(ret.k[:], raw, mem.B(nil)); err != nil {
+		return NodePublic{}, err
+	}
+	return ret, nil
+}
+
+func NodePublicFromRaw32(raw mem.RO) NodePublic {
```

Comment:
So Untyped is hex-encoded and Raw is not? Is there a package level comment somewhere about naming conventions?

Also, a doc string for this function would be nice; it would be non-obvious to me w/o the implementation.